### PR TITLE
Eliminate regex used to determine if content from output buffer is a valid HTML document since done actions now used

### PR DESCRIPTION
**Pull request by pierlon**
## Summary

Optimizes the number of steps that may be needed to determine if the content from the output buffer is a valid match.

## Checklist

- [ ] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by schlessera** [comment]
1. Why do we have the condition in there that the document needs to contain a `<head>` tag? In other places of the code, we're already sanitizing the document to add the `<head>` tag and its children if they are missing... Why not just check for the `<html>` tag only?

2. Because of the above, if you have a (valid even) HTML document that does not contain a `<head>` tag, it scans the entire document in a very slow way only to fail at the end:

Image 2020-08-12 at 11 56 32 AM

AFAIK, this document could be transformed into AMP without issues by the plugin.

**Comment by westonruter** [comment]
It's a good point. I think the intention of checking for `head` is to verify the response is likely to be a valid HTML page that is suitable for making into an AMP page. Nevertheless, this is now pretty much redundant with the much faster check we have in https://github.com/ampproject/amp-wp/pull/5202. 

So I think we should perhaps eliminate this condition altogether.

File: includes/class-amp-theme-support.php [hunk 1]
```diff
@@ -1941,7 +1941,7 @@ public static function prepare_response( $response, $args = [] ) {
 		 * Abort if the response was not HTML. To be post-processed as an AMP page, the output-buffered document must
 		 * have the HTML mime type and it must start with <html> followed by <head> tag (with whitespace, doctype, and comments optionally interspersed).
 		 */
-		if ( Attribute::TYPE_HTML !== substr( AMP_HTTP::get_response_content_type(), 0, 9 ) || ! preg_match( '#^(?:<!.*?>|\s+)*<html.*?>(?:<!.*?>|\s+)*<head\b(.*?)>#is', $response ) ) {
+		if ( Attribute::TYPE_HTML !== substr( AMP_HTTP::get_response_content_type(), 0, 9 ) || ! preg_match( '#^(?:<!.*?>|\s+)*+<html.*?>(?:<!.*?>|\s+)*+<head\b(.*?)>#is', $response ) ) {
 			return $response;
 		}
```
Comment (westonruter) [inline includes/class-amp-theme-support.php]: Yes, now that we have the `$did_template_action` check above, this condition should never evaluate to true. So I think the whole `if` statement can be removed now as being redundant.

Comment (pierlon) [reply includes/class-amp-theme-support.php]: Yea that makes sense.

Comment (westonruter) [reply includes/class-amp-theme-support.php]: Pushing that up…

**Review by westonruter — APPROVED** [review]

**Comment by schlessera** [comment]
Makes sense to me, although I'm feeling a bit uneasy about the strong assumptions this contains. I can see things like customized landing pages not get the AMP treatment because they intentionally skip the `'wp_head'` action...

**Comment by westonruter** [comment]
I think that that is good because it is unlikely that such highly custom pages would be normal pages intended for as AMP in the first place. Ref #5202

**Comment by schlessera** [comment]
I was thinking more about very simplistic, manually edited HTML where the author would opt to just avoid all the crap that plugins inject into a page. That type of page would probably be straight-forward to transform and profit well from AMP support.

**Comment by westonruter** [comment]
Yeah, though they could write the AMP themselves in that case too. Or else, the CWV likely won't be so bad and AMP wouldn't even be so important. In any case, if the head and footer actions are absent then other aspects of the AMP plugin will break, like adding the amphtml link and mobile redirection. 

It is a trade off, but it seems to me that being conservative in this case is going to break less things. AMP is not required for a site to work. We should always expect that it can be turned off for any page. Turning it off (at least by default) for such custom templates should increase compatibility with the ecosystem (e.g. the Password Protected plugin).

**Comment by schlessera** [comment]
Alright, I'm sold! ;)

**Comment by schlessera** [comment]
**QA passed**

The regex check is indeed removed and the validation requests that are terminating early are not throwing any errors.

**Pull request merged**